### PR TITLE
Button: Getting anchor native props if href prop is specified

### PR DESCRIPTION
**Pull request by khmakoto**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes

This PR gets `anchorProperties` instead of `buttonProperties` if the `href` prop has been specified for the `Button` component. Examples and snapshot tests have been added for this scenario as part of this PR.

I also realized that a lot of snapshot tests were not running correctly because it was missing the call to the `toMatchSnapshot` function, so this PR also fixes that.

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9456)

### CODE REVIEW

File: packages/experiments/src/components/Button/Button.view.tsx [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
 export const ButtonView: IButtonComponent['view'] = props => {
   const { icon, content, children, disabled, onClick, allowDisabledFocus, ariaLabel, keytipProps, buttonRef, ...rest } = props;
 
+  const rootType = _deriveRootType(props);
+  const htmlType = rootType === 'a' ? 'link' : 'button';
+  const propertiesType = rootType === 'a' ? anchorProperties : buttonProperties;
+
   // TODO: 'href' is anchor property... consider getNativeProps by root type
-  const buttonProps = { ...getNativeProps(rest, buttonProperties) };
+  const buttonProps = { ...getNativeProps(rest, propertiesType) };
```
Comment (JasonGore) [inline packages/experiments/src/components/Button/Button.view.tsx]: TS3.5 is requiring explicit types for `getNativeProps` calls so you may have to add one here. You can search any other use of `getNativeProps` for examples.

File: packages/experiments/src/components/Button/Button.view.tsx [hunk 2]
```diff
@@ -1,19 +1,23 @@
...
 import { Icon } from '../../utilities/factoryComponents';
 
 import { IButtonComponent, IButtonProps, IButtonRootElements, IButtonSlots, IButtonViewProps } from './Button.types';
 
 export const ButtonView: IButtonComponent['view'] = props => {
   const { icon, content, children, disabled, onClick, allowDisabledFocus, ariaLabel, keytipProps, buttonRef, ...rest } = props;
 
+  const rootType = _deriveRootType(props);
+  const htmlType = rootType === 'a' ? 'link' : 'button';
+  const propertiesType = rootType === 'a' ? anchorProperties : buttonProperties;
```
Comment (joschect) [inline packages/experiments/src/components/Button/Button.view.tsx]: This is starting to make me feel that this view is doing too much and button vs anchor should be split out into smaller components. That way you can check once if there is an href and then have each component handle it's own props correctly.

Comment (joschect) [reply packages/experiments/src/components/Button/Button.view.tsx]: Or maybe there should be a function like _deriveRootState or something that bundles all that work into one place.

Comment (khmakoto) [reply packages/experiments/src/components/Button/Button.view.tsx]: I was thinking about the second option you mentioned yesterday and I think I'm going to go ahead and combine those 3 lines into one function.

Comment (khmakoto) [reply packages/experiments/src/components/Button/Button.view.tsx]: @joschect Updated with the function approach bundling that work 👍

File: packages/experiments/src/components/Button/Button.view.tsx (hunk 1 above)
Comment (KevinTCoughlin) [reply packages/experiments/src/components/Button/Button.view.tsx]: Here's an example: https://github.com/OfficeDev/office-ui-fabric-react/blob/6ff7873e60436513ce7c3e4f39982a461fa5c86f/packages/office-ui-fabric-react/src/components/Persona/PersonaCoin/PersonaCoin.base.tsx#L72

Just swap the HTMLElement typing.

**Pull request merged**